tests/components/generic_thermostat/test_climate.py

```diff
@@ -763,102 +764,126 @@ async def setup_comp_4(hass):
+    _setup_sensor(hass, sensor_temperature)
     await hass.async_block_till_done()
     assert len(calls) == 0
 
 
-async def test_temp_change_ac_trigger_off_long_enough(
-    hass: HomeAssistant, setup_comp_4
+@pytest.mark.parametrize(
+    (
+        "ac_mode",
+        "initial_hvac_mode",
+        "initial_switch_state",
+        "sensor_temperature",
+        "target_temperature",
+        "triggered_service_call",
+    ),
+    [
+        (True, HVACMode.COOL, False, 30, 25, SERVICE_TURN_ON),
+        (True, HVACMode.COOL, True, 25, 30, SERVICE_TURN_OFF),
+        (False, HVACMode.HEAT, False, 25, 30, SERVICE_TURN_ON),
+        (False, HVACMode.HEAT, True, 30, 25, SERVICE_TURN_OFF),
+    ],
+)
+async def test_heating_cooling_switch_toggles_when_outside_min_cycle_duration(
+    hass: HomeAssistant,
+    ac_mode: bool,
+    initial_hvac_mode: HVACMode,
+    initial_switch_state: bool,
+    sensor_temperature: int,
+    target_temperature: int,
+    triggered_service_call: str,
 ) -> None:
-    """Test if temperature change turn ac on."""
+    """Test if heating/cooling toggles when outside minimum cycle."""
+    await _setup_thermostat_with_min_cycle_duration(hass, ac_mode, initial_hvac_mode)
     fake_changed = datetime.datetime(1970, 11, 11, 11, 11, 11, tzinfo=dt_util.UTC)
     with freeze_time(fake_changed):
-        calls = _setup_switch(hass, True)
-    await common.async_set_temperature(hass, 30)
-    _setup_sensor(hass, 25)
+        calls = _setup_switch(hass, initial_switch_state)
+    await common.async_set_temperature(hass, target_temperature)
+    _setup_sensor(hass, sensor_temperature)
     await hass.async_block_till_done()
     assert len(calls) == 1
     call = calls[0]
     assert call.domain == HASS_DOMAIN
-    assert call.service == SERVICE_TURN_OFF
+    assert call.service == triggered_service_call
     assert call.data["entity_id"] == ENT_SWITCH
 
 
-async def test_mode_change_ac_trigger_off_not_long_enough(
-    hass: HomeAssistant, setup_comp_4
-) -> None:
-    """Test if mode change turns ac off despite minimum cycle."""
-    calls = _setup_switch(hass, True)
-    await common.async_set_temperature(hass, 30)
-    _setup_sensor(hass, 25)
-    await hass.async_block_till_done()
-    assert len(calls) == 0
-    await common.async_set_hvac_mode(hass, HVACMode.OFF)
-    assert len(calls) == 1
-    call = calls[0]
-    assert call.domain == "homeassistant"
-    assert call.service == SERVICE_TURN_OFF
-    assert call.data["entity_id"] == ENT_SWITCH
-
-
-async def test_mode_change_ac_trigger_on_not_long_enough(
-    hass: HomeAssistant, setup_comp_4
+@pytest.mark.parametrize(
+    (
+        "ac_mode",
+        "initial_hvac_mode",
+        "initial_switch_state",
+        "sensor_temperature",
+        "target_temperature",
+        "changed_hvac_mode",
+        "triggered_service_call",
+    ),
+    [
+        (True, HVACMode.COOL, False, 30, 25, HVACMode.HEAT, SERVICE_TURN_ON),
```

Comment:
Using `HAVCMode.HEAT` does not seem right here with an air conditioning mode, but it is actually what the original test code does:

https://github.com/home-assistant/core/blob/0548f9f3425e65449ff2d0eabe76d99b83bc715b/tests/components/generic_thermostat/test_climate.py#L857

Is this a bug? Should this be changed?

Reply:
As a note: I also tested with `changed_hvac_mode = HVACMode.COOL` and the test passes. Awaiting feedback to know how to deal with this situation.

Reply:
The problem is the default implementation of the `turn_on` service. In a separate PR, we should make it possible to add proper turn_on and turn_off to generic thermostat, and maybe only set `ClimateEntityFeature.TURN_ON` + `ClimateEntityFeature.TURN_OFF` if that's enabled.

More details in https://github.com/home-assistant/core/issues/109195